src/python/grpcio_observability/grpc_observability/_observability.py

```diff
@@ -0,0 +1,172 @@
+    grpc.StatusCode.PERMISSION_DENIED: "PERMISSION_DENIED",
+    grpc.StatusCode.UNAUTHENTICATED: "UNAUTHENTICATED",
+    grpc.StatusCode.RESOURCE_EXHAUSTED: "RESOURCE_EXHAUSTED",
+    grpc.StatusCode.FAILED_PRECONDITION: "FAILED_PRECONDITION",
+    grpc.StatusCode.ABORTED: "ABORTED",
+    grpc.StatusCode.OUT_OF_RANGE: "OUT_OF_RANGE",
+    grpc.StatusCode.UNIMPLEMENTED: "UNIMPLEMENTED",
+    grpc.StatusCode.INTERNAL: "INTERNAL",
+    grpc.StatusCode.UNAVAILABLE: "UNAVAILABLE",
+    grpc.StatusCode.DATA_LOSS: "DATA_LOSS",
+}
+
+
+@dataclass
+class GcpObservabilityPythonConfig:
+    _singleton = None
+    _lock: threading.RLock = threading.RLock()
+    project_id: str = ""
+    stats_enabled: bool = False
+    tracing_enabled: bool = False
+    labels: Optional[Mapping[str, str]] = field(default_factory=dict)
+    sampling_rate: Optional[float] = 0.0
+
+    @staticmethod
+    def get():
+        with GcpObservabilityPythonConfig._lock:
+            if GcpObservabilityPythonConfig._singleton is None:
+                GcpObservabilityPythonConfig._singleton = GcpObservabilityPythonConfig(
+                )
+        return GcpObservabilityPythonConfig._singleton
+
+    def set_configuration(self,
+                          project_id: str,
+                          sampling_rate: Optional[float] = 0.0,
+                          labels: Optional[Mapping[str, str]] = None,
+                          tracing_enabled: bool = False,
+                          stats_enabled: bool = False) -> None:
+        self.project_id = project_id
+        self.stats_enabled = stats_enabled
+        self.tracing_enabled = tracing_enabled
+        self.labels = labels
+        self.sampling_rate = sampling_rate
+
+
+# pylint: disable=no-self-use
+class GCPOpenCensusObservability(grpc._observability.ObservabilityPlugin):
+    config: GcpObservabilityPythonConfig
+    exporter: "grpc_observability.Exporter"
+
+    def __init__(self, exporter: "grpc_observability.Exporter" = None):
+        self.exporter = None
+        self.config = GcpObservabilityPythonConfig.get()
+        if exporter:
+            self.exporter = exporter
+        else:
+            pass
+            # Actual implementation of OC exporter
+            # open_census = importlib.import_module(
+            #     "grpc_observability._open_census")
+            # self.exporter = open_census.OpenCensusExporter(
+            #     self.config.get().labels)
+        config_valid = _cyobservability.set_gcp_observability_config(
+            self.config)
+        if not config_valid:
+            raise ValueError("Invalid configuration")
+
+        if self.config.tracing_enabled:
+            self.set_tracing(True)
+        if self.config.stats_enabled:
+            self.set_stats(True)
+
+    def exit(self) -> None:
+        # Sleep for 0.5s so all data can be flushed.
+        # The time equals to the time in AwaitNextBatchLocked.
+        time.sleep(_cyobservability.CENSUS_EXPORT_BATCH_INTERVAL)
```

Comment:
> The sleep is considered a kludge. We can add some synchronization with the export thread to exit as soon as all data has been exported.
* It's not enough to exit as soon as all data has been exported, If we shutdown export thread immediately after exit, it's possible that core didn't call RecordEnd in callTracer, and all data recorded by calling RecordEnd will be lost.
* It's also possible to ignore those metrics since we have already exited the observability context, so anything recorded after this can be ignored. However, I think we should keep it because RecordEnd is part of RPC, and we should at least try our best (sleep one interval) to make sure it's called before exit.

Reply:
> It's also possible to ignore those metrics since we have already exited the observability context, so anything recorded after this can be ignored. However, I think we should keep it because RecordEnd is part of RPC, and we should at least try our best 

This seems reasonable. The guarantee would be that "any RPC started while the context manager is active would be recorded".

So the concern is a situation like this?

- RPC A starts
- `GCPOpenCensusObservability.exit()` is called
  - A flush of the buffer is initiated
  - The export of the census data finishes
- `GCPOpenCensusObservability.exit()` returns based on its synchronization with the export thread
- RPC A finishes and `StartEnd` is called, but o11y has finished

I can see this being an issue. Could you just keep track of the number of outstanding RPCs and keep stats collection open until either all of the RPCs have finished or a timeout has elapsed? (You don't want to just block indefinitely for RPCs to finish)

> If we shutdown export thread immediately after exit, it's possible that core didn't call RecordEnd in callTracer

My concern that led to the request for synchronization was actually the opposite -- that by waiting for too short a period of time, the stats would not be exported in time, the process would exit, and the buffered data would be lost.

This relates to the lifecycle of the stats exporting thread. I just realized that this is not currently a daemon thread, so if the process tries to exit, it will block waiting for the stats thread to exit. This could happen if, for example, there is a bug in the opencensus library, so I _do_ think we should switch to a daemon thread.

But that introduces the problem with the process exiting (and killing the exporting thread) before the stats have been exported. My thought was that the `exit()` method here would block until either exporting has finished or we hit a user-configurable timeout.